x-pack/plugins/upgrade_assistant/server/lib/reindexing/reindex_service.ts

```diff
@@ -0,0 +1,498 @@
+   * @param reindexOp
+   */
+  const startReindexing = async (reindexOp: ReindexSavedObject) => {
+    const { indexName } = reindexOp.attributes;
+    const reindexBody = {
+      source: { index: indexName },
+      dest: { index: reindexOp.attributes.newIndexName },
+    } as any;
+
+    const booleanFieldPaths = await actions.getBooleanFieldPaths(indexName);
+    if (booleanFieldPaths.length) {
+      reindexBody.script = {
+        lang: 'painless',
+        source: `
+          // Updates a single field in a Map
+          void updateField(Map data, String fieldName) {
+            if (
+              data[fieldName] == 'yes' ||
+              data[fieldName] == '1' ||
+              (data[fieldName] instanceof Integer && data[fieldName] == 1) ||
+              data[fieldName] == 'on'
+            ) {
+              data[fieldName] = true;
+            } else if (
+              data[fieldName] == 'no' ||
+              data[fieldName] == '0' ||
+              (data[fieldName] instanceof Integer && data[fieldName] == 0) ||
+              data[fieldName] == 'off'
+            ) {
+              data[fieldName] = false;
+            }
+          }
+
+          // Recursively walks the fieldPath list and calls
+          void updateFieldPath(def data, List fieldPath) {
+            String pathHead = fieldPath[0];
+
+            if (fieldPath.getLength() == 1) {
+              if (data.get(pathHead) !== null) {
+                updateField(data, pathHead);
+              }
+            } else {
+              List fieldPathTail = fieldPath.subList(1, fieldPath.getLength());
+
+              if (data.get(pathHead) instanceof List) {
+                for (item in data[pathHead]) {
+                  updateFieldPath(item, fieldPathTail);
+                }
+              } else if (data.get(pathHead) instanceof Map) {
+                updateFieldPath(data[pathHead], fieldPathTail);
+              }
+            }
+          }
+
+          for (fieldPath in params.booleanFieldPaths) {
+            updateFieldPath(ctx._source, fieldPath)
+          }
+        `,
+        params: { booleanFieldPaths },
+      };
+    }
+
+    const startReindex = (await callCluster('reindex', {
+      refresh: true,
+      waitForCompletion: false,
+      body: reindexBody,
+    })) as { task: string };
+
+    return actions.updateReindexOp(reindexOp, {
+      lastCompletedStep: ReindexStep.reindexStarted,
+      reindexTaskId: startReindex.task,
+      reindexTaskPercComplete: 0,
+    });
+  };
+
+  /**
+   * Polls Elasticsearch's Tasks API to see if the reindex operation has been completed.
+   * @param reindexOp
+   */
+  const updateReindexStatus = async (reindexOp: ReindexSavedObject) => {
+    const taskId = reindexOp.attributes.reindexTaskId;
+
+    // Check reindexing task progress
+    const taskResponse = await callCluster('tasks.get', {
+      taskId,
+      waitForCompletion: false,
+    });
+
+    if (taskResponse.completed) {
```

Comment:
Do you think we should validate the document counts? Or is the task completing without any errors sufficient?

Reply:
Something different than the check on the next line?

Reply:
Yeah, it would be comparing the document counts against the current source index. 

One way I could see this happening:

- someone presses the button to re-index
- it breaks their app because it was read-only
- they re-enable writes on the index
- re-index completes and the index is swapped, resulting in data loss

Not sure if this is necessary, but since we're essentially deleting an index I want to make sure we're covering all angles.

Reply:
Got it, I'll do a count request on the existing index and compare to the number of docs created by the reindex.